### PR TITLE
A range of refreshments ...

### PR DESCRIPTION
**Pull request by kernc**
fixes https://github.com/DmitryUlyanov/Multicore-TSNE/issues/27
fixes https://github.com/DmitryUlyanov/Multicore-TSNE/issues/24
fixes https://github.com/DmitryUlyanov/Multicore-TSNE/issues/14
fixes https://github.com/DmitryUlyanov/Multicore-TSNE/issues/13
closes https://github.com/DmitryUlyanov/Multicore-TSNE/pull/16

### CODE REVIEW

**Comment by kernc** [comment]
Ah, sorry, I have not intended to open a PR here just yet, instead wanted to test it first here https://github.com/biolab/Multicore-TSNE/pull/1.

The proposed changes include fixes for a couple of reported bugs / enhancement requests (fixes https://github.com/DmitryUlyanov/Multicore-TSNE/issues/27, fixes, https://github.com/DmitryUlyanov/Multicore-TSNE/issues/24, fixes https://github.com/DmitryUlyanov/Multicore-TSNE/issues/14, fixes https://github.com/DmitryUlyanov/Multicore-TSNE/issues/13), and also include the rebased verbosity patches provided by @bkj in https://github.com/DmitryUlyanov/Multicore-TSNE/pull/16, as well as some other healthy or needed extensions.

I'm eager to get your opinion on all these. Thanks for a great, performant library!

**Comment by DmitryUlyanov** [comment]
Hi! That's pretty big PR, notify when you are ready to merge and I will review it!

**Comment by kernc** [comment]
I think it's fairly complete. The [tests pass](https://travis-ci.org/biolab/Multicore-TSNE/builds/290422225). You might consider enabling Travis for this repo. I tried to make commits as disjunct as reasonably possible, and I already merged all the fixup commits into their parents, so there are no "fix typo from previous commit" here. I think it's good for a preliminary review. 🤞

**Comment by kernc** [comment]
What I meant with disjunct commits is that it might be best to review each commit separately.

File: multicore_tsne/tsne.cpp [hunk 1]
```diff
@@ -491,18 +491,8 @@ void TSNE::zeroMean(double* X, int N, int D) {
...
-        x = 2 * (rand() / ((double) RAND_MAX + 1)) - 1;
-        y = 2 * (rand() / ((double) RAND_MAX + 1)) - 1;
-        radius = (x * x) + (y * y);
-    } while ((radius >= 1.0) || (radius == 0.0));
-    radius = sqrt(-2 * log(radius) / radius);
-    x *= radius;
-    y *= radius;
-    return x;
+double TSNE::random() {
+    return rand() / ((double) RAND_MAX + 1);
```
Comment (DmitryUlyanov) [inline multicore_tsne/tsne.cpp]: Why did you decide to switch from normal to uniform RV?

**Comment by DmitryUlyanov** [comment]
That is a very solid PR! Especially thanks for improving installation process and in particular `setup.py`. I don't have any more comments than the one about uniform/normal noise. 

I will check that tests work and merge it.

File: multicore_tsne/tsne.cpp (hunk 1 above)
Comment (kernc) [reply multicore_tsne/tsne.cpp]: Decide? You give me too much credit. 😄 I just tried to take every possible further optimization step I could find. And this function loops and calls sqrt and log; what appears to be, observing the results on the few datasets I ran it with, completely redundant.

But seeing now how simplification adds no real speed advantage, I'm happy to purge it.

**Comment by kernc** [comment]
Thanks. I have rebased the branch to exclude the uniform random commit.

**Comment by DmitryUlyanov** [comment]
Thanks once again!

**Comment by DmitryUlyanov** [comment]
Oh, forgot one thing. Is installation process for Mac written in README is now obsolete? Does it compile normally?  Can you please update the instructions if necessary?

**Pull request merged**